CoinList 화면 콜렉션 뷰 구현

z3rosmith opened the pull request:
- Diffable datasource, CollectionViewListCell을 이용해서 collectionview를 구성하였습니다
- CoinListDataSource 파일을 만들어서 diffable datasourc로 사용하였습니다
- Xib를 이용해서 collection view list cell을 커스텀하였습니다.

<img width="564" alt="스크린샷 2022-02-26 오전 12 03 15" src="https://user-images.githubusercontent.com/52317025/155737835-e80993dc-915a-456b-bb4a-8eda5da1479a.png">

z3rosmith committed:
feat: CoinList화면 구현
- Diffable datasource, CollectionViewListCell을 이용해서
  collectionview를 구성하였습니다
- CoinListDataSource 파일을 만들어서 diffable datasourc로
  사용하였습니다
- Xib를 이용해서 collection view list cell을 커스텀하였습니다.

z3rosmith committed:
style: 마크 주석 추가
- 마크 주석 추가
- 공백 조정

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListDataSource.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+class CoinListDataSource {
+    struct Coin: Hashable {
```
CoinListViewController 타입 내부 중첩 타입에는 `// MARK: - Nested Type`가 달려있는데 여기도 추가되면 좋을 것 같습니다:)

donggeonoh replied:
저도 허황의 의견에 동참합니다 ㅎㅎ
근데 전에 작업했던 입출금 현황 화면은 Temporary Type 으로 한것 같기두 ㅎㅎ..

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+import UIKit
+
+class CoinListViewController: UIViewController {
+    // MARK: - Nested Type
```
스타일 쉐어 컨벤션을 주석 파트 부분 가져왔습니다:)
// MARK:를 사용해서 연관된 코드를 구분짓습니다.

Objective-C에서 제공하는 #pragma mark와 같은 기능으로, 연관된 코드와 그렇지 않은 코드를 구분할 때 사용합니다.

```swift
// MARK: Init

override init(frame: CGRect) {
  // doSomething()
}

deinit {
  // doSomething()
}


// MARK: Layout

override func layoutSubviews() {
  // doSomething()
}


// MARK: Actions

override func menuButtonDidTap() {
  // doSomething()
}
```

MARK 주석 위로는 두줄, 아래로는 한줄 띄워져 있습니다:)

z3rosmith replied:
넵 반영하도록 하겠습니다!

다만

스타일쉐어에 

### 빈 줄
- 빈 줄에는 공백이 포함되지 않도록 합니다.
- 모든 파일은 빈 줄로 끝나도록 합니다.
- MARK 구문 위와 아래에는 공백이 필요합니다.

```swift

// MARK: Layout

override func layoutSubviews() {
  // doSomething()
}

// MARK: Actions

override func menuButtonDidTap() {
  // doSomething()
}
```

이런 내용도 있는데 윗쪽으로도 한칸만 띄우는거는 어떻게 생각하세요?

donggeonoh replied:
이부분 저는 붙이는게 개인적인 스타일이긴 한데, 스타일쉐어를 따를까요? 전 따르는게 좋다고 생각이 드네요~

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+// MARK: - Configuration
+extension CoinListViewController {
+    func configureDataSource() {
+        let coinCellRegistration = UICollectionView.CellRegistration<CoinListCollectionViewCell, CoinListDataSource.Coin>(cellNib: UINib(nibName: "CoinListCollectionViewCell", bundle: nil)) { cell, indexPath, item in
```
UINib(nibName: "CoinListCollectionViewCell", bundle: nil)를 별도의 상수에 담아서 넣어주는 것도 좋을 것 같습니다.
코드가 너무 길어서 가독성이 조금 떨어지는 것 같아요 ㅜㅜ
또, 컨벤션 확인 부탁드립니다:)

z3rosmith replied:
말씀해주신대로 UINib 생성 부분은 따로 빼도록 하겠습니다!

그런데 혹시 컨벤션은 어떤부분을 말씀하시는건가요?

donggeonoh replied:
허황 말대로 이부분 제 코드에서도 수정 해야겠네요 ㅠㅠ

HJEHA replied:
코드가 너무 길어서 줄바꿈이 필요해 보이는데 회의실에서 의견 나눠도 괜찮을까요?:)

donggeonoh replied:
좋습니다~

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+extension CoinListViewController {
+    func configureDataSource() {
+        let coinCellRegistration = UICollectionView.CellRegistration<CoinListCollectionViewCell, CoinListDataSource.Coin>(cellNib: UINib(nibName: "CoinListCollectionViewCell", bundle: nil)) { cell, indexPath, item in
+            cell.nameLabel.text = item.callingName
+            cell.symbolPerCurrencyLabel.text = item.symbolName + "/KRW"
+            cell.coinPriceLabel.text = "\(item.price)"
+            cell.changeRateLabel.text = "+ \(item.changeRate)%"
+            cell.changePriceLabel.text = "+ \(item.changePrice)"
```
셀의 프로퍼티에 직접 접근하기보다 item을 매개변수로 받는 메서드를 만드는건 어떻게 생각하시나요???:)

z3rosmith replied:
어떤 형태를 말씀하시는지 이해가 잘 안 되는데 혹시 구체적으로 설명해주실 수 있으실까요?

donggeonoh replied:
저도 동감합니다. ViewCell의 프로퍼티는 private로 하고, 메서드를 만들어 매개변수로 타입을 받아서 ViewCell에서 값을 세팅해주면 좋을 것 같습니다.

HJEHA replied:
아래와 같은 메서드가 셀 xib에 연결된 클래스에 있는건 어떤가 말씀드렸던건데 설명이 부족했네요ㅜㅜ

```swift
func updateCell(to item: Item) {
    nameLabel.text = item.callingName
    symbolPerCurrencyLabel.text = item.symbolName + "/KRW"
    coinPriceLabel.text = "\(item.price)"
    changeRateLabel.text = "+ \(item.changeRate)%"
    changePriceLabel.text = "+ \(item.changePrice)"
}
```

HJEHA reviewed (commented):
코멘트 남겼습니다:)
확인 부탁드립니다!

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataSource.swift:
```diff
@@ -0,0 +1,27 @@
+        let changePrice: Int
+        let identifier = UUID()
+        
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(identifier)
+        }
```
UUID를 사용하셨군요. 좋은 것 같네요 ㅎㅎ
제가 구현한 입출금 현황 화면의 모델같은 경우 hashable만 채택했네요 ㅎㅎ
만약, 모든 프로퍼티의 값이 같다면 저같은 경우 오류가 발생 하겠지만, UUID를 사용하면 그런 경우가 발생하지 않겠네요!
좋은 정보 알아갑니다~

z3rosmith replied:
사실 애플 공식문서도 그렇고 Diffable DataSource에서는 다 UUID를 사용하는거같아요😊

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+
+class CoinListViewController: UIViewController {
+    // MARK: - Nested Type
+    enum Section: CaseIterable {
```
혹시 Section에 CaseIterable을 채택한 이유가 무엇일까요?

z3rosmith replied:
headerRegistration에서 section header의 이름을 지정해주기 위해 사용합니다!

```swift
let headerRegistration = UICollectionView.SupplementaryRegistration<UICollectionViewListCell>(elementKind: UICollectionView.elementKindSectionHeader) { headerView, elementKind, indexPath in
    var configuration = headerView.defaultContentConfiguration()
    configuration.text = Section.allCases[indexPath.section].description
    configuration.textProperties.font = .preferredFont(forTextStyle: .largeTitle)
    configuration.textProperties.color = .label
    headerView.contentConfiguration = configuration
}
```

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    }
+    
+    // MARK: - IBOutlet
+    @IBOutlet weak var coinListCollectionView: UICollectionView!
```
다른 곳에서 사용중이지 않는다면 private 접근제어자를 붙여주는게 좋다고 생각이드네요~

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    @IBOutlet weak var coinListCollectionView: UICollectionView!
+    
+    // MARK: - Property
+    var dataSource: UICollectionViewDiffableDataSource<Section, CoinListDataSource.Coin>!
```
이부분 혹시 var라고 한 이유가 있을까요? 아니면 private var가 맞다고 생각이드네요. 그리고 강제적으로 옵셔널을 언래핑해주고 있는데, 옵셔널로 가져가도 문제 없을 것 같습니다.

z3rosmith replied:
반영하겠습니다~!

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+            return collectionView.dequeueConfiguredReusableCell(using: coinCellRegistration, for: indexPath, item: item)
+        }
+        let headerRegistration = UICollectionView.SupplementaryRegistration<UICollectionViewListCell>(elementKind: UICollectionView.elementKindSectionHeader) { headerView, elementKind, indexPath in
+            var configuration = headerView.defaultContentConfiguration()
+            configuration.text = Section.allCases[indexPath.section].description
+            configuration.textProperties.font = .preferredFont(forTextStyle: .largeTitle)
+            configuration.textProperties.color = .label
```
이 부분 메서드화 하는 것도 좋을 것 같습니다.
이 의견은 꼭 반영 안되어도 좋습니다~

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+        dataSource = UICollectionViewDiffableDataSource<Section, CoinListDataSource.Coin>(collectionView: coinListCollectionView) { collectionView, indexPath, item in
+            return collectionView.dequeueConfiguredReusableCell(using: coinCellRegistration, for: indexPath, item: item)
+        }
+        let headerRegistration = UICollectionView.SupplementaryRegistration<UICollectionViewListCell>(elementKind: UICollectionView.elementKindSectionHeader) { headerView, elementKind, indexPath in
```
개인적으로 궁금해서 질문 드립니다~
혹시 UICollectionView.elementKindSectionHeader 가 어떤 역할을 하는지 여쭤봐도 될까요?

z3rosmith replied:
supplementaryView를 콜렉션뷰에서 그려줄 때 identifier처럼 사용되는것으로 알고있습니다!

donggeonoh replied:
감사합니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/View/CoinListView/CoinListCollectionViewCell.swift:
```diff
@@ -7,11 +7,15 @@
-class CoinListCollectionViewCell: UICollectionViewCell {
+class CoinListCollectionViewCell: UICollectionViewListCell {
 
+    @IBOutlet weak var nameLabel: UILabel!
+    @IBOutlet weak var symbolPerCurrencyLabel: UILabel!
+    @IBOutlet weak var coinPriceLabel: UILabel!
+    @IBOutlet weak var changeRateLabel: UILabel!
+    @IBOutlet weak var changePriceLabel: UILabel!
```
위에서 언급했지만 private 으로 하고 메서드로 모델을 넘겨받아서 수정하는 것이 좋아 보입니다~

donggeonoh reviewed (changes requested):
접근 제어자 정도만 수정하시면 될 것 같네요!
궁금한 점도 같이 코멘트 남깁니다 ㅎㅎ

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    func configureDataSource() {
+        let coinCellRegistration = UICollectionView.CellRegistration<CoinListCollectionViewCell, CoinListDataSource.Coin>(cellNib: UINib(nibName: "CoinListCollectionViewCell", bundle: nil)) { cell, indexPath, item in
+            cell.nameLabel.text = item.callingName
+            cell.symbolPerCurrencyLabel.text = item.symbolName + "/KRW"
```
symbolPerCurrencyLabel.text에 item.symbolName + "/KRW" 넣어줄 때 문자열을 `+` 연산자를 사용하셨는데 이유가 있으신가요??

`Coin struct 타입` 내부에 `symbolPerCurrency` 같은 연산프로퍼티를 만들고 그 연산 프로퍼티를 Label.text에 할당하는건 어떻게 생각하시나요??:)

z3rosmith committed:
refactor: PR 건의사항 반영#15
- 타입 내에서만 사용되는 프로퍼티는 private으로 변경
- cell의 외부에서 직접 프로퍼티 접근보다는 cell.update로 변경
- MARK 주석 추가
- dataSource !옵셔널에서 ?옵셔널로 변경
